### ISSUE TITLE
(Preview) TwitchChatClient & PubSub - requested disconnect, clients are reconnecting

### ISSUE THREAD

**Issue by WrithemTwine** [TwitchLib/TwitchLib#1093]
I know these are preview versions, was trying them based on some issues getting resolved.

When recently updated:
TwitchLib.Client - 3.2.6-preview-efac467b4b08e2ed974742f775517361850ffe33
TwitchLib.PubSub - 3.2.5-preview-e27b78571148879a04eaea2cf45441a35b0e908f

I've noticed an issue when I initiate a disconnect, with both the chat client and pub sub.

The Chat Client logs the disconnect and then a few seconds later begins the connection again. I have an event handler for "OnDisconnect" in case the user doesn't initiate a disconnect for a reconnection and I have a flag (IsStarted) to ensure a reconnection doesn't occur. A log of the action:
![image](https://user-images.githubusercontent.com/20450441/175756752-536220de-792f-4be1-8466-df8362776bb7.png)

The PubSub, sent a disconnect, and the logs show a response with an http 400 error continuously afterward, instead of straight up disconnecting the connection. I may have fixed this with a stop listening to a topic method call.

I know it's preview, but just trying to find a stable accurate version as TwitchChatClient 3.2.4 had join channel issues. The repo (with this issue) is currently at: https://github.com/WrithemTwine/TwineChatBot, https://github.com/WrithemTwine/TwineChatBot/archive/refs/tags/v.0.2.1.0-alpha_streamer.zip, under the "StreamerBot" project, with Chat Client programmed at: https://github.com/WrithemTwine/TwineChatBot/blob/main/StreamerBotLib/BotClients/Twitch/TwitchBotChatClient.cs. I'm hoping to issue a v.0.2.2.0 soon no longer reflecting these TwitchLib versions.

I appreciate all of your hardwork efforts.

**Comment by Mahsaap** [comment]
If you don't want client to reconnect I think you need to set max attempts in the reconnectpolicy in client options.

As for pubsub im not sure about that one , I'm assuming your hooked to onpubsubservicedisconnect to ensure it disconnect before you try to do soemthing else with the instance?

**Comment by WrithemTwine** [comment]
chat client, at least 2 scenarios exist when stopping the client:
- user initiated stop - I think the client should not try to reconnect - my bot code (tries to) gives the user an option to disconnect when stream is offline (stops ping pongs for 1+ days), or they can leave it running to handle commands & events.
- issue causing client disconnect (like an ISP lag, or something connection disruption, or the like) - I think the client should reconnect to maintain service, especially during a stream. Or does this mean, my code should handle/check all disconnects and reconnect attempts via _set max attempts in the reconnectpolicy in client options?_

Should a "hard-stop ( Disconnect( ) ) to the chat client" stop it?
Some of my workarounds involved creating a new chat client (as trying to restart often failed, and joining channels was an issue) when starting the bot (especially after a stop then another start), then the code ended up receiving 2+ duplicated messages since the code event handlers were registered to the new instance and the orphaned instance-that wasn't destroyed/garbage collected.


For the pubsub, I'm reading the output log and the responses start as regular responses, then sent 'disconnect()', and the responses become "bad requests"
![image](https://user-images.githubusercontent.com/20450441/175787900-099edc7d-e605-44bd-91c1-5b7307c7d43e.png)

And likewise with the chat client, "Disconnect()" doesn't "disconnect" the pubsub service. I believe when I debugged stopping PubSub with Disconnect(), the "OnDisconnect" and "OnConnect" event handlers both fired.

I have no issues once the behavior is finally determined-can code to that, it's just that connect/disconnect meanings are now changed (disconnect() is no longer really disconnect() without some other setting) and affects code flow diagrams. 

I really do appreciate a resilient service, just some of the naming conventions portray a behavior different than anticipated - leading to confusion about what really drives the procedure.

**Comment by Mahsaap** [comment]
Ya I dont think this is intended. There have been some connection issues lately with people not reconnecting, so maybe some of this is an after effect. I'll tag @Syzuna that would know more about connection related code.

**Comment by WrithemTwine** [comment]
no worries @Mahsaap, a lot going on everywhere.
I had other code to write-up anyway.

Thanks for the help.

**Comment by D1360-64RC14** [comment]
Holy... I took 2 days debugging my code to understand that `TwitchClient.Disconnect()` was reconnecting... wtf.

I'm trying to reconnect with other credentials when the token get expired.

```xml
<PackageReference Include="TwitchLib.Client" Version="3.3.1" />
```

**Comment by WrithemTwine** [comment]
yep. I was hoping they'd address, "reconnect - yes reconnect from an issue", "disconnect() - stop reconnecting-I want it off"; scenario. It didn't seem to have made it in latest release. I just disconnect the event handlers and start another bot instance to work around this until it might be better addressed.

**Comment by Syzuna** [comment]
could you try the preview version "3.4.0-preview-a14cf3a5db58c361ff356fbfa4a929a522fc64d7" and see if that small change in the bundled TwitchLib.Communication package fixed your issue?

**Comment by WrithemTwine** [comment]
Per my testing, calling Disconnect() resulted in a subsequent connect.

I tried: 
![image](https://user-images.githubusercontent.com/20450441/221018000-502d3718-06eb-4a45-880f-010f2ad5a61e.png)
and
![image](https://user-images.githubusercontent.com/20450441/221017480-3ee8f796-feeb-4e3e-a299-34cdedd034a8.png)

Both gave this similar output:
![image](https://user-images.githubusercontent.com/20450441/221017095-fdbf6ea7-23cb-46c3-881e-0b2ad8600b93.png)

Code calling Disconnect(): 
(I have 2 code versions with preprocessor directives. One version uses the same instance for connect/disconnect--the code issue I tested [some of the preprocessor directive limited code changed and needed updating-no more 'beinghosted' checks and 'RefreshSettings()' is called but no longer used in my code path]. The other version starts a new instance each connect/disconnect cycle.) https://github.com/WrithemTwine/TwineChatBot/blob/main/StreamerBotLib/BotClients/Twitch/TwitchBotChatClient.cs#L227

and 'ClientOptions' with ReconnectionPolicy

ClientOptions options = new()
            {
                UseSsl = true,
                ClientType = ClientType.Chat,
                MessagesAllowedInPeriod = TwitchClientID == TwitchChannelName ? 100 : 20,
                ThrottlingPeriod = TimeSpan.FromSeconds(30),
                SendQueueCapacity = 100,
                SendDelay = 5,
                WhisperQueueCapacity = 200,
                WhispersAllowedInPeriod = 200,
                WhisperThrottlingPeriod = TimeSpan.FromSeconds(30),
                ReconnectionPolicy = new(30)
            };


I believe I'm using the library correctly to connect the chat client, then disconnect the client. 
To summarize earlier posts, I give the option with my bot to connect the chat client only when the stream is live and disconnect after stream ends, or a user can keep the chat client online all the time. Hence, disconnect is somewhat important so a user can stop client ping/pongs between streams.

**Comment by Syzuna** [comment]
I've just pushed a new TwitchLib.Client preview version (3.4.0-preview-b78278fa1c106621cfecac2c52cdbd2ddef19143) with another potential fix (thanks to @christian-riedel )
Could you give it a try?

**Comment by WrithemTwine** [comment]
cool!

![image](https://user-images.githubusercontent.com/20450441/221270483-0005bba4-098d-4163-b9f9-cf4b3de65d24.png)

TwitchChatClient hasn't tried to reconnect in over 2 minutes (compared to 5 second reconnect in previous test).

**Comment by WrithemTwine** [comment]
It then seems to get stuck during Connect() <-fine, Disconnect() <- fine, and Connect() <-stuck reconnecting to channel

![image](https://user-images.githubusercontent.com/20450441/221271508-af96276f-ad87-4237-aef2-3667e91b5669.png)

**Comment by christian-riedel** [comment]
@Syzuna and @all you're welcome


@WrithemTwine

> It then seems to get stuck during Connect() <-fine, Disconnect() <- fine, and Connect() <-stuck reconnecting to channel
> 
> ![image](https://user-images.githubusercontent.com/20450441/221271636-44750ccb-2ca9-4f31-924f-998db1d0b851.png)

i also experienced that issue/behaviour

first at all, there is still a bug in the code/fix that i provided

the order/sequence or the called methods itself are important

whenever the `TwichtClient` has been connected and disconnected manually once, to reconnect to twitch a call to `TwitchClient.Reconnect()` has to be made

something to keep in mind: after disconnecting manually, channels have to be joined manually again via `TwitchClient.JoinChannel(...)`

**Comment by WrithemTwine** [comment]
Thanks for helping! 
interesting twist. I don't see an issue as it's a different code path for reconnecting, just needs to be clear when using the code.
I won't be able to test for several hours at least.

**Comment by WrithemTwine** [comment]
I'll wait until these issues settle and the preview releases get finalized.
I'll have to program 'first start' and 'subsequent start' code paths since that determines whether Twitch.Connect() or Twitch.Reconnect() are used, and whether the JoinChannels call is necessary.

**Comment by Bukk94** [comment]
Manually joining using `JoinChannel` sounds really counterintuitive. I would expect the same behavior when `Connect` is called. no matter if it's first or second call after `Disconnect()`.
Just wondering, wouldn't be simpler to convert those clients into proper async? This is like a timebomb waiting to explode and I bet it's hard to track those loose Tasks. 
But I'm so glad that someone had the courage to finally take a look at this because the code is huge mess.

**Comment by WrithemTwine** [comment]
Yes, I agree:
"specify join channel->Connect()->use it", then
"specify join channel->Connect()->Disconnect()->Reconnect()->specify join channel"
seems out of ordinary repetitive.

From what I've seen in the code, the joined channels object gets cleared as each channel is added as a chat client. Not saving them seems odd and removes information. The Reconnect() is unable to then recreate user provided channel information, instead of the user managing the existing channels. Having the channel strings saved in the client object doesn't seem to cost that much memory, even a comma delimited string as storage, when compared to being more respondent to the user. This might alleviate more code issues.

As well, when connected, keeping the tasks helps keep them from floating anonymously in memory. Be sure completed tasks are disposed, or else there could be a memory leak/overflow (the garbage collector doesn't reclaim them). (I once had a method that kept cycling through Tasks and never left current scope, so the memory overflowed after a lot of tasks completed but didn't get disposed).

I've noticed taking out private class data, or working around to minimize code paths, often make it harder to provide a responsive outward interface. I've made a point to stay out of the TwitchLib library (code contribution) to keep it separate from my own bot code.

**Comment by christian-riedel** [comment]
ok. i give it a try and started rewriting

my approach is to start with the Communication-Lib, to continue with the Client-Lib and to go over to the PuSub-Lib

for those who are interested, the ongoing process is available/can be found over there

https://github.com/christian-riedel/TwitchLib.Communication/tree/dev
https://github.com/christian-riedel/TwitchLib.Client/tree/dev
https://github.com/christian-riedel/TwitchLib.PubSub/tree/dev
